### PR TITLE
ProcessGroupUCC tests

### PR DESCRIPTION
**Pull request by zasdfgbnm**
- [x] Direct dependency on UCX is completely removed, UCC active set API always enabled
- [x] Remove `TORCH_UCC_PROFILING_ENABLE`, always enable profiling
- [x] Fixes profiling of `recv` and `all_gather`
- [x] Use the NCCL TL of UCC on CUDA, as  the UCP TL is not well supported on CUDA

Most tests are passing, but there are a few skipped tests:
- `scatter` and `gather` are not supported by the UCP TL of UCC on CPU tensors
- A few flaky tests in PyTorch's CI environment
- Profiler-related failures, some of them will be fixed by @Fuzzkatt in https://github.com/pytorch/pytorch/pull/84368

After this PR is merged, I will continue to work on these skipped failures.

### CODE REVIEW

File: test/run_test.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 import sys
 import tempfile
 import json
+#from turtle import back
```
Comment (zasdfgbnm) [inline test/run_test.py:17]: TOTO: remove before commit

File: test/run_test.py [hunk 2]
```diff
@@ -493,6 +501,8 @@ def test_distributed(test_module, test_directory, options):
         print_to_stderr("MPI not available -- MPI backend tests will be skipped")
     config = DISTRIBUTED_TESTS_CONFIG
     for backend, env_vars in config.items():
+        if backend != "ucc":
+            continue
```
Comment (zasdfgbnm) [inline test/run_test.py:505]: TODO: remove before commit

**Comment by zasdfgbnm** [comment]
@vtlam @kwen2501 This PR is ready for review. Unfortunately, many DDP tests are still hanging, so I have to skip them in this PR, and I will open a tracking issue for failing tests. This PR is already big enough, so fixing of unit tests should go to separate PR in the future.

File: test/run_test.py [hunk 3]
```diff
@@ -298,6 +298,12 @@ def skip_test_p(name: str) -> bool:
             "WORLD_SIZE": "2" if torch.cuda.device_count() == 2 else "3",
             "TEST_REPORT_SOURCE_OVERRIDE": "dist-gloo",
         }
+    if dist.is_ucc_available():
+        DISTRIBUTED_TESTS_CONFIG["ucc"] = {
+            "WORLD_SIZE": "2" if torch.cuda.device_count() == 2 else "3",
+            "TEST_REPORT_SOURCE_OVERRIDE": "dist-ucc",
+            "UCX_TLS": "tcp,cuda",
```
Comment (kingchc) [inline test/run_test.py:305]: `cuda` protocol will likely result in deadlock (i.e., timeout) in UCC PG due to the multi-thread limitation of CUDA (i.e., main thread and progress thread in UCC PG cannot perform CUDA calls will cause deadlock). I would recommend either using NCCL TL for cuda tests and/or use just `UCX_TLS =tcp` for UCP TL

Comment (zasdfgbnm) [reply test/run_test.py]: done

File: torch/testing/_internal/distributed/distributed_test.py [hunk 4]
```diff
@@ -143,20 +143,23 @@ def eq(value, other):
...
     dist.Backend.NCCL,
+    # TODO(ucc): dist.Backend.UCC,
 ]
 
 # Allowlist of distributed backends where profiling is supported for p2p ops
 SEND_RECV_PROFILING_SUPPORTED_BACKENDS = [
     dist.Backend.MPI,
     dist.Backend.GLOO,
     dist.Backend.NCCL,
+    # TODO(ucc): dist.Backend.UCC,
```
Comment (zasdfgbnm) [inline torch/testing/_internal/distributed/distributed_test.py:162]: These failures are fixed by @Fuzzkatt in https://github.com/pytorch/pytorch/pull/84368

File: torch/testing/_internal/distributed/distributed_test.py [hunk 5]
```diff
@@ -9018,6 +9033,7 @@ def _test_hook_pickling(self, hook, hook_state):
             BACKEND not in DistTestCases.backend_feature["cuda"],
             f"The {BACKEND} backend does not support DDP communication hook on CUDA devices"
         )
+        @sandcastle_skip_if(BACKEND == "ucc", "TODO(ucc): investigate why failing")
```
Comment (zasdfgbnm) [inline torch/testing/_internal/distributed/distributed_test.py:9036]: This test is flaky in the CI environment:
```
No such file or directory: ‘/tmp/checkpoint.pt’
```
It is not a problem of the UCC PG

File: test/run_test.py [hunk 6]
```diff
@@ -298,6 +298,14 @@ def skip_test_p(name: str) -> bool:
             "WORLD_SIZE": "2" if torch.cuda.device_count() == 2 else "3",
             "TEST_REPORT_SOURCE_OVERRIDE": "dist-gloo",
         }
+    if dist.is_ucc_available():
+        DISTRIBUTED_TESTS_CONFIG["ucc"] = {
+            "WORLD_SIZE": "2" if torch.cuda.device_count() == 2 else "3",
+            "TEST_REPORT_SOURCE_OVERRIDE": "dist-ucc",
+            "UCX_TLS": "tcp",
+            "UCC_TLS": "nccl,ucp",
+            "UCC_TL_UCP_TUNE": "cuda:0",
```
Comment (vtlam) [inline test/run_test.py:307]: Add a comment that this rule implies UCP TL is not used for GPU tensors.

Comment (zasdfgbnm) [reply test/run_test.py]: added

**Comment by kumpera** [comment]
Is this PR ready to be merged? It's passing all tests.

**Review by vtlam — APPROVED** [review]
LGTM. Thanks Xiang! @kwen2501 @malfet could you take a look? Thanks!

File: torch/csrc/distributed/c10d/ProcessGroup.hpp [hunk 7]
```diff
@@ -411,7 +414,7 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
       int /* tag */) {
     TORCH_CHECK(
         false,
-        c10::str("ProcessGroup ", getBackendName(), "does not support send"));
+        c10::str("ProcessGroup ", getBackendName(), " does not support send"));
```
Comment (malfet) [inline torch/csrc/distributed/c10d/ProcessGroup.hpp:417]: Absolutely love the changes, but do you mind moving them to a separate PR (so that we can land it quickly and reduce size of this one)

**Review by malfet — APPROVED** [review]
LGTM, though can you please comment how much it will increase distributed tests runtime?

File: torch/csrc/distributed/c10d/ProcessGroup.hpp (hunk 7 above)
Comment (zasdfgbnm) [reply torch/csrc/distributed/c10d/ProcessGroup.hpp:417]: Split out at https://github.com/pytorch/pytorch/pull/84780

File: torch/testing/_internal/distributed/distributed_test.py [hunk 8]
```diff
@@ -2913,6 +2918,7 @@ def _test_scatter_helper(
             self._barrier()
 
         @sandcastle_skip_if(BACKEND == "nccl", "Nccl does not support CPU tensors")
+        @sandcastle_skip_if(BACKEND == "ucc", "CPU tensor ops not supported by UCP TL")
```
Comment (kwen2501) [inline torch/testing/_internal/distributed/distributed_test.py:2921]: Curious: what would happen if a user indeed makes a scatter call with CPU tensors?

**Review by kwen2501 — APPROVED** [review]
LGTM. Added a question.

**Comment by zasdfgbnm** [comment]
> pull / linux-bionic-cuda11.6-py3.10-gcc7 / test

@malfet I am comparing the `pull / linux-bionic-cuda11.6-py3.10-gcc7 / test` distributed workflows. In this PR, the three shards takes `1h 16m 57s`, `59m 17s`, `1h 3m 0s`, in https://github.com/pytorch/pytorch/pull/84780, the numbers are `1h 17m 14s`, `1h 1m 27s`, `1h 4m 33s`, so there CI time should be negligible.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 8 above)
Comment (zasdfgbnm) [reply torch/testing/_internal/distributed/distributed_test.py:2921]: User will get an error, for example:
```
RuntimeError: [/var/lib/jenkins/workspace/torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:463] [Rank 1][ProcessGroupUCC-1][READY]failed to init collective, error code -1: Operation is not supported, system error code 11
Exception raised from enqueue_collective at /var/lib/jenkins/workspace/torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:463 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7ff89899a60b in /opt/conda/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xce (0x7ff898995fde in /opt/conda/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #2: <unknown function> + 0x1d9148 (0x7ff8ae10c148 in /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cuda_cpp.so)
frame #3: c10d::ProcessGroupUCC::scatter(std::vector<at::Tensor, std::allocator<at::Tensor> >&, std::vector<std::vector<at::Tensor, std::allocator<at::Tensor> >, std::allocator<std::vector<at::Tensor, std::allocator<at::Tensor> > > >&, c10d::ScatterOptions const&) + 0x77b (0x7ff8ae11f5ab in /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cuda_cpp.so)
frame #4: <unknown function> + 0x4548189 (0x7ff8a6647189 in /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
frame #5: <unknown function> + 0x454da04 (0x7ff8a664ca04 in /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
frame #6: <unknown function> + 0x4559348 (0x7ff8a6658348 in /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
frame #7: c10d::ops::scatter(c10::intrusive_ptr<c10d::ProcessGroup, c10::detail::intrusive_target_default_null_type<c10d::ProcessGroup> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, std::vector<std::vector<at::Tensor, std::allocator<at::Tensor> >, std::allocator<std::vector<at::Tensor, std::allocator<at::Tensor> > > > const&, c10d::ScatterOptions const&) + 0x2ac (0x7ff8a664b51c in /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
frame #8: <unknown function> + 0x9b21cd (0x7ff8b048e1cd in /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_python.so)
frame #9: <unknown function> + 0x242487 (0x7ff8afd1e487 in /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_python.so)
frame #10: <unknown function> + 0x17ffa4 (0x55cc596f2fa4 in /opt/conda/bin/python)
frame #11: _PyObject_MakeTpCall + 0x15e (0x55cc596b552e in /opt/conda/bin/python)
frame #12: <unknown function> + 0xaea74 (0x55cc59621a74 in /opt/conda/bin/python)
frame #13: <unknown function> + 0x1052a5 (0x55cc596782a5 in /opt/conda/bin/python)
frame #14: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #15: _PyObject_Call + 0xf7 (0x55cc596bae97 in /opt/conda/bin/python)
frame #16: _PyEval_EvalFrameDefault + 0x26e4 (0x55cc59756774 in /opt/conda/bin/python)
frame #17: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #18: <unknown function> + 0x104425 (0x55cc59677425 in /opt/conda/bin/python)
frame #19: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #20: <unknown function> + 0x104425 (0x55cc59677425 in /opt/conda/bin/python)
frame #21: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #22: <unknown function> + 0x18fc9b (0x55cc59702c9b in /opt/conda/bin/python)
frame #23: <unknown function> + 0x1050b5 (0x55cc596780b5 in /opt/conda/bin/python)
frame #24: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #25: <unknown function> + 0x1049f3 (0x55cc596779f3 in /opt/conda/bin/python)
frame #26: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #27: <unknown function> + 0x18fc9b (0x55cc59702c9b in /opt/conda/bin/python)
frame #28: <unknown function> + 0x104425 (0x55cc59677425 in /opt/conda/bin/python)
frame #29: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #30: <unknown function> + 0x18fc9b (0x55cc59702c9b in /opt/conda/bin/python)
frame #31: <unknown function> + 0x104425 (0x55cc59677425 in /opt/conda/bin/python)
frame #32: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #33: <unknown function> + 0x1049f3 (0x55cc596779f3 in /opt/conda/bin/python)
frame #34: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #35: <unknown function> + 0x18fd4e (0x55cc59702d4e in /opt/conda/bin/python)
frame #36: _PyObject_Call + 0x20a (0x55cc596bafaa in /opt/conda/bin/python)
frame #37: _PyEval_EvalFrameDefault + 0x26e4 (0x55cc59756774 in /opt/conda/bin/python)
frame #38: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #39: <unknown function> + 0x1049f3 (0x55cc596779f3 in /opt/conda/bin/python)
frame #40: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #41: <unknown function> + 0x1049f3 (0x55cc596779f3 in /opt/conda/bin/python)
frame #42: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #43: <unknown function> + 0x104425 (0x55cc59677425 in /opt/conda/bin/python)
frame #44: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #45: <unknown function> + 0x1050b5 (0x55cc596780b5 in /opt/conda/bin/python)
frame #46: <unknown function> + 0x18f742 (0x55cc59702742 in /opt/conda/bin/python)
frame #47: PyEval_EvalCode + 0x88 (0x55cc59711578 in /opt/conda/bin/python)
frame #48: <unknown function> + 0x2548fb (0x55cc597c78fb in /opt/conda/bin/python)
frame #49: <unknown function> + 0x28d3a5 (0x55cc598003a5 in /opt/conda/bin/python)
frame #50: PyRun_StringFlags + 0x7a (0x55cc598051ca in /opt/conda/bin/python)
frame #51: PyRun_SimpleStringFlags + 0x3c (0x55cc5980522c in /opt/conda/bin/python)
frame #52: Py_RunMain + 0x46d (0x55cc598056cd in /opt/conda/bin/python)
frame #53: Py_BytesMain + 0x39 (0x55cc598057d9 in /opt/conda/bin/python)
frame #54: __libc_start_main + 0xe7 (0x7ff8e7d9ec87 in /lib/x86_64-linux-gnu/libc.so.6)
frame #55: <unknown function> + 0x2125d4 (0x55cc597855d4 in /opt/conda/bin/python)
```
